server/views/pages/visitors.html

```diff
@@ -0,0 +1,124 @@
+{% from "../components/page-navigation/macro.njk" import pageNavigation %}
+{% from "../components/user-details/macro.njk" import userDetails %}
+{% from "../components/feedback-widget/macro.njk" import hubFeedbackWidget %}
+
+
+{% extends "../components/template.njk" %}
+
+{% block pageTitle %}
+  {{ title }}
+{% endblock %}
+
+{% block head %}
+  <link href="/public/stylesheets/application.css" rel="stylesheet"/>
+{% endblock %}
+
+{% block pageNavigation %}
+  {{ pageNavigation({ title: title }) }}
+{% endblock %}
+
+{% block topBar %}
+  {{ topBar({
+    establishmentDisplayName: establishmentDisplayName,
+    showBar: false,
+    userName: signedInUser,
+    returnUrl: returnUrl
+  }) }}
+{% endblock %}
+
+{% block userDetails %}
+  {{ userDetails({
+    detailsType: detailsType,
+    userName: signedInUser
+  }) }}
+{% endblock %}
+
+{% block header %}
+  <div class="govuk-width-container govuk-body">
+    <h1>{{ title }}</h1>
+  </div>
+{% endblock %}
+
+{% block homeNavigation %}{% endblock %}
+{% block search %}{% endblock %}
+
+{% block main %}
+  <div class="govuk-body">
+    {% if not signedInUser %}
+      <div class= "govuk-width-container">
+        <h2 class="govuk-heading-m">You are signed out</h2>
+        <p class="govuk-!-font-size-24" data-test="signin-prompt">
+          <a href="/auth/sign-in?returnUrl={{ returnUrl }}" class="govuk-link">Sign in</a> to see your personal information.
+          </p>
+      </div>
+    {% else %}
+      <div class= "govuk-width-container  govuk-!-padding-bottom-9">
+        {% set endVisitor = currentVisitorsPage*maxVisitorsPerPage %}
+        {% set startVisitor = endVisitor-maxVisitorsPerPage %}
+        {% set numberOfPages = (approvedVisitors.length/maxVisitorsPerPage) | round(0, "ceil") %}
+        {% if endVisitor > approvedVisitors.length %}
+          {% set endVisitor = approvedVisitors.length %}
+        {% endif %}
+        <dl class="govuk-summary-list">
+          {% for i in range(startVisitor, endVisitor) %}
+            <div class="govuk-summary-list__row">
+              <dd class="govuk-summary-list__value">
+                {{ approvedVisitors[i] }}
+              </dd>
+            </div>
+          {% endfor %}
+        </dl>
+        {% if numberOfPages > 1 %}
+          {% set pageItems = [] %}
+          {% for i in range(1, numberOfPages+1) %}
+            {% set pageItems = (pageItems.push({ text: i, href: '?page='+i, selected: i==currentVisitorsPage }), pageItems) %}
```

Comment:
I'd have been tempted to push all the pagination logic into a util thing server side and unit test it and then just have this simply reading data.

we use this kind of approach in dps: 
https://github.com/ministryofjustice/use-of-force/blob/c43ee10d24335f3e0eb7523648a7fe24bf418211/server/utils/page.ts

And an associated `toPagination` [filter](https://github.com/ministryofjustice/use-of-force/blob/351987c2bbe9867f11890f647203d3f8f6b3f158/server/utils/nunjucksSetup.ts#L124), that takes the metadata and generates the correct info for the page controls

Reply:
Agreed, I prefer to use filters for stuff like this. The project wasn't using them, but now the floodgates are open, I think we need to readdress other areas. I'm sure we've got some tech debt where filters would be more appropriate.